File: CS50x/sentimental-credit/credit.py
import re


def main():
    # Regual expression patters for each card
    regex_numeric = r"\d+"
    regex_master = r"5[1-5]\d{14}"
    regex_visa = r"4\d{12}(\d{3})?"
    regex_amex = r"3[47]\d{13}"
    # Make sure to get a valid credit card number (numeric)
    while True:
        num = input("Credit card number: ")
        if re.fullmatch(regex_numeric, num):
            break

    # Find out which kind of card it is
    if re.fullmatch(regex_master, num):
        card = "MASTERCARD"
    elif re.fullmatch(regex_visa, num):
        card = "VISA"
    elif re.fullmatch(regex_amex, num):
        card = "AMEX"
    else:
        card = "INVALID"

    # Check if valid
    if checksum(int(num)):
        print(card)
    else:
        print("INVALID")

# Check luhn_algorithm


def checksum(n):
    # Find the sum of its digits (by an algorithm)
    sum = 0
    while n > 0:
        sum += n % 10
        n = n // 10
        sum += dig_sum((n % 10) * 2)
        n = n // 10

    if sum % 10 == 0:
        return True

    return False


def dig_sum(i):
    sum = 0
    while i > 0:
        sum += i % 10
        i = i // 10
    return sum


if __name__ == "__main__":
    main()

File: CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// functions
int sum_dig(int i);
int check_luhn(long n);
int num_of_dig(long number);
int fst2_dig(long number);
void card_check(long num);

// main
int main(void)
{
    // get a credit card number
    long c_num = get_long("Number: ");

    // check Luhn’s Sum
    if (!check_luhn(c_num))
    {
        card_check(c_num);
    }
    else
    {
        printf("INVALID\n");
    }
}

int check_luhn(long n)
{
    // variable for the doubled sum
    int double_sum = 0;

    // variable for the regular sum
    int reg_sum = 0;

    // find the sums
    while (n)
    {
        reg_sum += n % 10;
        n = n / 10;
        double_sum += sum_dig((n % 10) * 2);
        n = n / 10;
    }

    // checks whether it is valid or not
    return ((double_sum + reg_sum) % 10);
}

int sum_dig(int i)
{
    // check whether it is 1 or 2 digit number
    if (i < 10)
    {
        return i;
    }
    else
    {
        return ((i / 10) + (i % 10));
    }
}

void card_check(long num)
{
    // variables with needed atributes of a number
    int number_of_digits = num_of_dig(num);
    int first_digits = fst2_dig(num);

    // find which card it is
    if ((first_digits == 34 || first_digits == 37) && number_of_digits == 15)
    {
        printf("AMEX\n");
    }
    else if ((first_digits > 50 && first_digits < 56) && number_of_digits == 16)
    {
        printf("MASTERCARD\n");
    }
    else if ((first_digits / 10 == 4) && (number_of_digits == 16 || number_of_digits == 13))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int num_of_dig(long number)
{
    // check the number of digits
    int counter = 0;
    while (number > 0)
    {
        number = number / 10;
        counter++;
    }
    return counter;
}

int fst2_dig(long number)
{
    // find the first 2 digits of a number
    while (number > 99)
    {
        number = number / 10;
    }
    return number;
}
